### ISSUE TITLE
Gallery format: enhance captions to be adaptable beyond the mere property value

### ISSUE THREAD

**Issue by the1gofer** [SemanticMediaWiki/SemanticResultFormats#376]
It would be nice if could have a little more control over the captions that are displayed, maybe with a template or something similar. 

For example:
```
==Census Pages==
{{#ask: [[has source::United States Census, 1870]]
|?has person
|format=gallery 
|captionproperty=has person
}}
```

assuming the document has ` [[has person::Bob Smith]]`. 

The caption will read "Bob Smith".  I would like to be able to do something like "The household of Bob Smith" or the "Households of Bob Smith, John Doe" (if their were more than one person).   I feel like this could be done with a template, but I'm not an expert,and could be wrong.   

It also seems like it would be able to have better control over where that links.  Following the documentation example provided makes the link to "bob smith"  I would just like to link back to the image.

**Comment by mwjames** [comment]
> (if their were more than one person).   I feel like this could be done with
> a template, but I'm not an expert,and could be wrong.
>
> It also seems like it would be able to have better control over where that
> links.  Following the documentation example provided makes the link to "bob
> smith"  I would just like to link back to the image.

I see your use case but the gallery class isn't easily extendable but
for anyone who wants to give it a shot, the following can be used as
blueprint:

```
@@ -312,10 +312,16 @@ class Gallery extends ResultPrinter {
 		} else {
 			if ( $imgTitle instanceof Title && $imgTitle->getNamespace() ==
NS_FILE && !$this->isSpecialPage() ) {
 				$imgCaption = $ig->mParser->recursiveTagParse( $imgCaption );
 			}
 		}
+
+		if ( $this->m_params['templatecaptions'] !== '' ) {
+			$templateCode = "{{" . $this->m_params['templatecaptions'] .
"|imagecaption=$imgCaption|imageredirect=$imgRedirect}}";
+			$imgCaption = wfMessage( 'smw-parse', $templateCode )->parse();
+		}
+
 		// Use image alt as helper for either text
 		$imgAlt =  $this->params['redirects'] === '' ? $imgCaption :
$imgRedirect !== '' ? $imgRedirect : '' ;
 		$ig->add( $imgTitle, $imgCaption, $imgAlt );
 	}

@@ -456,10 +462,16 @@ class Gallery extends ResultPrinter {
 			'type' => 'boolean',
 			'default' => true,
 			'message' => 'srf_paramdesc_autocaptions'
 		];

+		$params['templatecaptions'] = [
+			'type' => 'string',
+			'default' => '',
+			'message' => 'srf-paramdesc-templatecaptions'
+		];
+
```

PS: The code isn't tested and may need modifications (as said it is a
blueprint!).

Cheers

**Comment by kghbln** [comment]
Thanks for making this suggestion to enhance this format.

**Comment by gesinn-it** [comment]
closed via #626
